src/chain/shared_state.rs

```diff
@@ -104,116 +96,30 @@ impl SharedState {
-            log_ident,
-            "handled_genesis_event",
-            &self.handled_genesis_event,
-            &false,
-        );
-        self.handled_genesis_event = true;
-
-        if related_info.is_empty() {
-            return Ok(());
+    pub fn update(&mut self, new: Option<Self>, log_ident: &LogIdent) {
+        if self.handled_genesis_event {
+            log_or_panic!(
+                log::Level::Error,
+                "{} - shared state update - genesis event already handled",
+                log_ident
+            );
         }
 
-        let (
-            our_infos,
-            our_history,
-            our_members,
-            neighbour_infos,
-            their_keys,
-            their_knowledge,
-            their_recent_keys,
-            churn_event_backlog,
-            relocate_queue,
-        ) = deserialize(related_info)?;
-        if self.our_infos.len() != 1 {
-            // Check nodes with a history before genesis match the genesis block:
-            update_with_genesis_related_info_check_same(
-                log_ident,
-                "our_infos",
-                &self.our_infos,
-                &our_infos,
-            );
-            update_with_genesis_related_info_check_same(
-                log_ident,
-                "our_history",
-                &self.our_history,
-                &our_history,
-            );
-            update_with_genesis_related_info_check_same(
-                log_ident,
-                "our_members",
-                &self.our_members,
-                &our_members,
-            );
-            update_with_genesis_related_info_check_same(
-                log_ident,
-                "neighbour_infos",
-                &self.neighbour_infos,
-                &neighbour_infos,
-            );
-            update_with_genesis_related_info_check_same(
-                log_ident,
-                "their_keys",
-                &self.their_keys,
-                &their_keys,
-            );
-            update_with_genesis_related_info_check_same(
-                log_ident,
-                "their_knowledge",
-                &self.their_knowledge,
-                &their_knowledge,
-            );
-            update_with_genesis_related_info_check_same(
-                log_ident,
-                "their_recent_keys",
-                &self.their_recent_keys,
-                &their_recent_keys,
-            );
-            update_with_genesis_related_info_check_same(
-                log_ident,
-                "churn_event_backlog",
-                &self.churn_event_backlog,
-                &churn_event_backlog,
-            );
-            update_with_genesis_related_info_check_same(
-                log_ident,
-                "relocate_queue",
-                &self.relocate_queue,
-                &relocate_queue,
-            );
+        if let Some(new) = new {
+            if self.our_infos.len() > 1 && *self != new {
```

Comment:
hmm...   so when `self == new`, the later on `self = new` will still be executed ?

Reply:
Yes, but this same logic which exists in fleming too. I think the idea is that if we are already elder, we don't expect the shared state that comes from the genesis event to be different from the shared state we already have. We panic in tests when this happens because it indicates a bug in the code, but in production we just apply the new shared state as we should always trust whatever comes out of parsec.